### ISSUE TITLE
Exception occurs when setting custom timezone on Postgres db w/ read-only connection

### ISSUE THREAD

**Issue by iwaffles** [metabase/metabase#1699]
Running a custom query `select * from orders;` returns the following:

<img width="569" alt="error" src="https://cloud.githubusercontent.com/assets/7333/12076852/bcad8ee6-b173-11e5-8334-7f993a5a6cda.png">

Anyone have any tips on how to fix? I've tried reinstalling, upgrading, restarting the instance and syncing the database.

**Comment by salsakran** [comment]
Could you give us a bit of information about the DB you're using and what version of Metabase you're on? If you're comfortable sharing the schema publicly that would also be very useful.

Additionally, the server logs might have a more specific error message, do you have access to them by any chance?

**Comment by iwaffles** [comment]
Hey @salsakran sure thing!

Metabase 0.13.2 using PostgreSQL.

Unfortunately I cannot share the db schema publicly.

I am seeing this in one of the error logs: 

```
2016-01-03 17:53:53,057 [DEBUG] metabase.db.internal :: DB CALL:  Database [:engine :details] (where (limit (select* ENTITY21373) 1) {:id database-id})
2016-01-03 17:53:53,060 [DEBUG] metabase.driver.generic-sql.native :: [32mUPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';[0m
2016-01-03 17:53:53,137 [ERROR] metabase.driver.generic-sql.native :: [31mFailed to set timezone: Batch entry 0 UPDATE pg_settings SET setting = 'US/Pacific' WHERE name ILIKE 'timezone' was aborted.  Call getNextException to see the cause.[0m
2016-01-03 17:53:53,137 [DEBUG] metabase.driver.generic-sql.native :: [32mselect * from orders;[0m
2016-01-03 17:53:53,140 [WARN ] com.mchange.v2.c3p0.impl.NewPooledConnection :: [c3p0] A PooledConnection that has already signalled a Connection error is still in use!
2016-01-03 17:53:53,140 [WARN ] com.mchange.v2.c3p0.impl.NewPooledConnection :: [c3p0] Another error has occurred [ org.postgresql.util.PSQLException: ERROR: current transaction is aborted, commands ignored until end of transaction block ] which will not be reported to listeners!
org.postgresql.util.PSQLException: ERROR: current transaction is aborted, commands ignored until end of transaction block
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2198)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1927)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:255)
	at org.postgresql.jdbc2.AbstractJdbc2Statement.execute(AbstractJdbc2Statement.java:561)
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeWithFlags(AbstractJdbc2Statement.java:419)
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeQuery(AbstractJdbc2Statement.java:304)
	at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeQuery(NewProxyPreparedStatement.java:76)
	at clojure.java.jdbc$db_query_with_resultset$run_query_with_params__1807.invoke(jdbc.clj:824)
	at clojure.java.jdbc$db_query_with_resultset.invoke(jdbc.clj:831)
	at clojure.java.jdbc$query.doInvoke(jdbc.clj:866)
	at clojure.lang.RestFn.invoke(RestFn.java:464)
	at metabase.driver.generic_sql.native$process_and_run$fn__21381.invoke(native.clj:43)
	at clojure.java.jdbc$db_transaction_STAR_.doInvoke(jdbc.clj:595)
	at clojure.lang.RestFn.invoke(RestFn.java:521)
	at clojure.java.jdbc$db_transaction_STAR_.doInvoke(jdbc.clj:611)
	at clojure.lang.RestFn.invoke(RestFn.java:425)
	at metabase.driver.generic_sql.native$process_and_run.invoke(native.clj:27)
	at clojure.lang.Var.invoke(Var.java:383)
	at metabase.driver$fn__16355$G__16148__16362.invoke(driver.clj:36)
	at clojure.core$partial$fn__4527.invoke(core.clj:2493)
	at metabase.driver.query_processor$wrap_guard_multiple_calls$fn__18008.invoke(query_processor.clj:249)
	at metabase.driver.query_processor$pre_log_query$fn__18001.invoke(query_processor.clj:239)
	at metabase.driver.query_processor.annotate$post_annotate$fn__17361.invoke(annotate.clj:263)
	at metabase.driver.query_processor$limit$fn__17992.invoke(query_processor.clj:219)
	at metabase.driver.query_processor$cumulative_sum$fn__17986.invoke(query_processor.clj:208)
	at metabase.driver.query_processor$pre_add_implicit_breakout_order_by$fn__17943.invoke(query_processor.clj:144)
	at metabase.driver.query_processor$pre_add_implicit_fields$fn__17910.invoke(query_processor.clj:128)
	at metabase.driver.query_processor$post_add_row_count_and_status$fn__17893.invoke(query_processor.clj:87)
	at metabase.driver.query_processor$pre_expand$fn__17889.invoke(query_processor.clj:74)
	at metabase.driver.query_processor$wrap_catch_exceptions$fn__17882.invoke(query_processor.clj:59)
	at metabase.driver.query_processor$process.invoke(query_processor.clj:291)
	at metabase.driver$process_query.invoke(driver.clj:350)
	at metabase.driver$dataset_query.invoke(driver.clj:390)
	at metabase.api.dataset$fn__18015$fn__18018.invoke(dataset.clj:27)
	at metabase.api.common.internal$wrap_catch_api_exceptions.invoke(internal.clj:216)
	at metabase.api.dataset$fn__18015.invoke(dataset.clj:20)
	at compojure.core$make_route$fn__1460.invoke(core.clj:130)
	at compojure.core$wrap_route_middleware$fn__1456.invoke(core.clj:118)
	at compojure.core$if_route$fn__1432.invoke(core.clj:41)
	at compojure.core$if_method$fn__1424.invoke(core.clj:27)
	at compojure.core$routing$fn__1467.invoke(core.clj:144)
	at clojure.core$some.invoke(core.clj:2570)
	at compojure.core$routing.doInvoke(core.clj:144)
	at clojure.lang.RestFn.applyTo(RestFn.java:139)
	at clojure.core$apply.invoke(core.clj:632)
	at compojure.core$routes$fn__1471.invoke(core.clj:149)
	at metabase.middleware$enforce_authentication$fn__19700.invoke(middleware.clj:82)
	at compojure.core$routing$fn__1467.invoke(core.clj:144)
	at clojure.core$some.invoke(core.clj:2570)
	at compojure.core$routing.doInvoke(core.clj:144)
	at clojure.lang.RestFn.invoke(RestFn.java:423)
	at metabase.api.routes$fn__19812.invoke(routes.clj:38)
	at compojure.core$if_context$fn__1489.invoke(core.clj:211)
	at compojure.core$routing$fn__1467.invoke(core.clj:144)
	at clojure.core$some.invoke(core.clj:2570)
	at compojure.core$routing.doInvoke(core.clj:144)
	at clojure.lang.RestFn.applyTo(RestFn.java:139)
	at clojure.core$apply.invoke(core.clj:632)
	at compojure.core$routes$fn__1471.invoke(core.clj:149)
	at compojure.core$routing$fn__1467.invoke(core.clj:144)
	at clojure.core$some.invoke(core.clj:2570)
	at compojure.core$routing.doInvoke(core.clj:144)
	at clojure.lang.RestFn.invoke(RestFn.java:423)
	at metabase.routes$fn__20686.invoke(routes.clj:23)
	at compojure.core$if_context$fn__1489.invoke(core.clj:211)
	at compojure.core$routing$fn__1467.invoke(core.clj:144)
	at clojure.core$some.invoke(core.clj:2570)
	at compojure.core$routing.doInvoke(core.clj:144)
	at clojure.lang.RestFn.applyTo(RestFn.java:139)
	at clojure.core$apply.invoke(core.clj:632)
	at compojure.core$routes$fn__1471.invoke(core.clj:149)
	at metabase.middleware$log_api_call$fn__19799.invoke(middleware.clj:302)
	at metabase.middleware$add_security_headers$fn__19752.invoke(middleware.clj:192)
	at metabase.middleware$format_response$fn__19764.invoke(middleware.clj:245)
	at ring.middleware.json$wrap_json_body$fn__20510.invoke(json.clj:44)
	at ring.middleware.json$wrap_json_response$fn__20527.invoke(json.clj:87)
	at ring.middleware.keyword_params$wrap_keyword_params$fn__20555.invoke(keyword_params.clj:32)
	at ring.middleware.params$wrap_params$fn__20577.invoke(params.clj:57)
	at metabase.middleware$bind_current_user$fn__19704.invoke(middleware.clj:96)
	at metabase.middleware$wrap_current_user_id$fn__19678.invoke(middleware.clj:64)
	at metabase.middleware$wrap_api_key$fn__19719.invoke(middleware.clj:107)
	at metabase.middleware$wrap_session_id$fn__19671.invoke(middleware.clj:55)
	at ring.middleware.cookies$wrap_cookies$fn__20455.invoke(cookies.clj:171)
	at ring.middleware.session$wrap_session$fn__20674.invoke(session.clj:85)
	at ring.middleware.gzip$wrap_gzip$fn__20484.invoke(gzip.clj:65)
	at ring.adapter.jetty$proxy_handler$fn__19912.invoke(jetty.clj:24)
	at ring.adapter.jetty.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:497)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
2016-01-03 17:53:53,146 [DEBUG] metabase.db.internal :: DB CALL:  Database [:name :id :engine :details] (where (limit (select* ENTITY17794) 1) {:id database-id})
2016-01-03 17:53:53,152 [ERROR] metabase.driver :: [31mQuery failure: ERROR: current transaction is aborted, commands ignored until end of transaction block[0m
2016-01-03 17:53:53,180 [DEBUG] metabase.models.interface :: DB CALL:  QueryExecution 552
2016-01-03 17:53:53,182 [DEBUG] metabase.middleware :: [32mPOST /api/dataset 200 (131 ms)[0m
2016-01-03 17:53:54,362 [DEBUG] metabase.middleware :: [34mGET /api/health [0m
2016-01-03 17:53:54,362 [DEBUG] metabase.middleware :: [32mGET /api/health 200 (0 ms)[0m
2016-01-03 17:53:56,974 [DEBUG] metabase.middleware :: [34mGET /api/setting [0m
2016-01-03 17:53:56,975 [DEBUG] metabase.db.internal :: DB CALL:  User [:id :email :date_joined :first_name :last_name :last_login :is_superuser :is_active :is_staff] (where (limit (select* ENTITY19706) 1) {:id current-user-id})
```

Does that help?

**Comment by tlrobinson** [comment]
@iwaffles Did SQL queries work at some point? If so did anything about your database or Metabase installation change recently?

**Comment by iwaffles** [comment]
@tlrobinson They did, we had a few running until I tried to create a new one. Not sure if they broke before I tried to create the new one or after – I didn't notice until attempting to create a new SQL query. 

I haven't tried changing the timezone. Is it possible that AWS EB did this somehow? Maybe it occurred during an update and I didn't notice.

**Comment by iwaffles** [comment]
Huzzah, thanks @tlrobinson! I unset the database in /admin/settings to "Database Default" and it seems to have fixed the issue!

:+1: :+1:

**Comment by tlrobinson** [comment]
Interesting. Can you find out what version of Postgres you're running?

@camsaul didn't you fix this in #1431?

**Comment by iwaffles** [comment]
It looks like 9.3.6 - here's the full output from select version();

PostgreSQL 9.3.6 on x86_64-unknown-linux-gnu, compiled by gcc (GCC) 4.8.2 20140120 (Red Hat 4.8.2-16), 64-bit

**Comment by agilliland** [comment]
It looks like your postgres db didn't actually have support for that timezone.  What's in your `pg_timezone_names` table?

http://www.postgresql.org/docs/9.2/static/view-pg-timezone-names.html

**Comment by iwaffles** [comment]
@agilliland running `select * from pg_timezone_names where name like '%Pacific%';` I see the following:


| name                 | abbrev | utc_offset                                         | is_dst | 
|----------------------|--------|----------------------------------------------------|--------| 
| US/Pacific-New       | PST    | 0 years 0 mons 0 days -8 hours 0 mins -0.00 secs   | false  | 
| US/Pacific           | PST    | 0 years 0 mons 0 days -8 hours 0 mins -0.00 secs   | false  | 
| Pacific/Marquesas    | MART   | 0 years 0 mons 0 days -9 hours -30 mins -0.00 secs | false  | 
| Pacific/Wake         | WAKT   | 0 years 0 mons 0 days 12 hours 0 mins 0.00 secs    | false  | 
| Pacific/Chatham      | CHADT  | 0 years 0 mons 0 days 13 hours 45 mins 0.00 secs   | true   | 
| Pacific/Chuuk        | CHUT   | 0 years 0 mons 0 days 10 hours 0 mins 0.00 secs    | false  | 
| Pacific/Fiji         | FJST   | 0 years 0 mons 0 days 13 hours 0 mins 0.00 secs    | true   | 
| Pacific/Saipan       | ChST   | 0 years 0 mons 0 days 10 hours 0 mins 0.00 secs    | false  | 
| Pacific/Ponape       | PONT   | 0 years 0 mons 0 days 11 hours 0 mins 0.00 secs    | false  | 
| Pacific/Kwajalein    | MHT    | 0 years 0 mons 0 days 12 hours 0 mins 0.00 secs    | false  | 
| Pacific/Tongatapu    | TOT    | 0 years 0 mons 0 days 13 hours 0 mins 0.00 secs    | false  | 
| Pacific/Kosrae       | KOST   | 0 years 0 mons 0 days 11 hours 0 mins 0.00 secs    | false  | 
| Pacific/Pitcairn     | PST    | 0 years 0 mons 0 days -8 hours 0 mins -0.00 secs   | false  | 
| Pacific/Guadalcanal  | SBT    | 0 years 0 mons 0 days 11 hours 0 mins 0.00 secs    | false  | 
| Pacific/Easter       | EAST   | 0 years 0 mons 0 days -5 hours 0 mins -0.00 secs   | false  | 
| Pacific/Guam         | ChST   | 0 years 0 mons 0 days 10 hours 0 mins 0.00 secs    | false  | 
| Pacific/Midway       | SST    | 0 years 0 mons 0 days -11 hours 0 mins -0.00 secs  | false  | 
| Pacific/Rarotonga    | CKT    | 0 years 0 mons 0 days -10 hours 0 mins -0.00 secs  | false  | 
| Pacific/Gambier      | GAMT   | 0 years 0 mons 0 days -9 hours 0 mins -0.00 secs   | false  | 
| Pacific/Niue         | NUT    | 0 years 0 mons 0 days -11 hours 0 mins -0.00 secs  | false  | 
| Pacific/Samoa        | SST    | 0 years 0 mons 0 days -11 hours 0 mins -0.00 secs  | false  | 
| Pacific/Bougainville | BST    | 0 years 0 mons 0 days 11 hours 0 mins 0.00 secs    | false  | 
| Pacific/Funafuti     | TVT    | 0 years 0 mons 0 days 12 hours 0 mins 0.00 secs    | false  | 
| Pacific/Efate        | VUT    | 0 years 0 mons 0 days 11 hours 0 mins 0.00 secs    | false  | 
| Pacific/Johnston     | HST    | 0 years 0 mons 0 days -10 hours 0 mins -0.00 secs  | false  | 
| Pacific/Yap          | CHUT   | 0 years 0 mons 0 days 10 hours 0 mins 0.00 secs    | false  | 
| Pacific/Pago_Pago    | SST    | 0 years 0 mons 0 days -11 hours 0 mins -0.00 secs  | false  | 
| Pacific/Pohnpei      | PONT   | 0 years 0 mons 0 days 11 hours 0 mins 0.00 secs    | false  | 
| Pacific/Port_Moresby | PGT    | 0 years 0 mons 0 days 10 hours 0 mins 0.00 secs    | false  | 
| Pacific/Majuro       | MHT    | 0 years 0 mons 0 days 12 hours 0 mins 0.00 secs    | false  | 
| Pacific/Palau        | PWT    | 0 years 0 mons 0 days 9 hours 0 mins 0.00 secs     | false  | 
| Pacific/Truk         | CHUT   | 0 years 0 mons 0 days 10 hours 0 mins 0.00 secs    | false  | 
| Pacific/Apia         | WSDT   | 0 years 0 mons 0 days 14 hours 0 mins 0.00 secs    | true   | 
| Pacific/Nauru        | NRT    | 0 years 0 mons 0 days 12 hours 0 mins 0.00 secs    | false  | 
| Pacific/Kiritimati   | LINT   | 0 years 0 mons 0 days 14 hours 0 mins 0.00 secs    | false  | 
| Pacific/Galapagos    | GALT   | 0 years 0 mons 0 days -6 hours 0 mins -0.00 secs   | false  | 
| Pacific/Noumea       | NCT    | 0 years 0 mons 0 days 11 hours 0 mins 0.00 secs    | false  | 
| Pacific/Enderbury    | PHOT   | 0 years 0 mons 0 days 13 hours 0 mins 0.00 secs    | false  | 
| Pacific/Honolulu     | HST    | 0 years 0 mons 0 days -10 hours 0 mins -0.00 secs  | false  | 
| Pacific/Tahiti       | TAHT   | 0 years 0 mons 0 days -10 hours 0 mins -0.00 secs  | false  | 
| Pacific/Tarawa       | GILT   | 0 years 0 mons 0 days 12 hours 0 mins 0.00 secs    | false  | 
| Pacific/Wallis       | WFT    | 0 years 0 mons 0 days 12 hours 0 mins 0.00 secs    | false  | 
| Pacific/Auckland     | NZDT   | 0 years 0 mons 0 days 13 hours 0 mins 0.00 secs    | true   | 
| Pacific/Fakaofo      | TKT    | 0 years 0 mons 0 days 13 hours 0 mins 0.00 secs    | false  | 
| Pacific/Norfolk      | NFT    | 0 years 0 mons 0 days 11 hours 30 mins 0.00 secs   | false  | 
| Canada/Pacific       | PST    | 0 years 0 mons 0 days -8 hours 0 mins -0.00 secs   | false  |

**Comment by seregine** [comment]
Also seeing this problem on Heroku Postgres (which uses Amazon RDS)
PostgreSQL 9.4.3 on x86_64-unknown-linux-gnu, compiled by gcc (Ubuntu/Linaro 4.6.3-1ubuntu5) 4.6.3, 64-bit

**Comment by hopsoft** [comment]
Setting the report timezone under `Admin Panel > Settings > General > Report Timezone` to _Database Default_ solved the problem for us when we switched Metabase to point to a new RDS replica db. Thanks @iwaffles

**Comment by tbeseda** [comment]
I also encountered this with custom SQL. We are running 0.13.2 on EB pointed at a Heroku "Follower" running pg 9.4.4.
Setting the Report Timezone to Database Default fixed the issue.

**Comment by agilliland** [comment]
Thanks for the reports on this one, it appears that this problem is specific to the way Metabase attempts to set the connection timezone on Postgres database connections.

Unfortunately it may turn out that this is a problem that's hard for us to rectify.  The postgres database driver doesn't seem to allow you to specify the timezone as part of the connection information, so we attempt to set it directly before we issue the query.  However, in the event that the database is a read-only connection it appears this fails because technically we are issuing a SQL update statement.

We'll have to look into this a bit more and see what's possible, but in the meantime the work around is as posted by several users, which is to leave the Metabase report timezone set to `Database Default`.

**Comment by camsaul** [comment]
It should still be running the query if it fails to set the timezone. The code is written to log the error if setting the timezone fails and run the query without setting it:

```clojure
(try (kdb/transaction (k/exec-raw [(sql/set-timezone-sql driver) [(driver/report-timezone)]])
                      (f))
     (catch Throwable e
       (log/error (u/format-color 'red "Failed to set timezone:\n%s"
                    (with-out-str (jdbc/print-sql-exception-chain e))))
       (f)))
```

**Comment by agilliland** [comment]
It looked like the connection was getting marked as invalid as soon as an exception was being thrown, so even though we catch the exception and try to move on it's still failing.

**Comment by idris** [comment]
I just had this problem where my custom queries were throwing these errors.  The solution of setting timezone to "database default" worked for me as well.

Something about my database DID change which caused the problem: I switched from querying off a main database to querying off of a read-only follower.  Hope that helps!

**Comment by nevir** [comment]
> Unfortunately it may turn out that this is a problem that's hard for us to rectify. The postgres database driver doesn't seem to allow you to specify the timezone as part of the connection information, so we attempt to set it directly before we issue the query. However, in the event that the database is a read-only connection it appears this fails because technically we are issuing a SQL update statement.

@agilliland from limited testing, it appears like `SET local timezone to 'America/Los_Angeles';` *does* work on a read-only replica

**Comment by camsaul** [comment]
It might be that your DB is already in pacific time so the SET operation doesn't do anything? Does it work with other time zones?

**Comment by pedro** [comment]
+1, `SET` should work on read-only dbs.

I tried patching metabase to use it, but couldn't get things to work... Probably doing something silly because I'm not familiar with Clojure, but this was my diff:

```diff
diff --git a/src/metabase/driver/postgres.clj b/src/metabase/driver/postgres.clj
index 5e3f7d6..b84d8d5 100644
--- a/src/metabase/driver/postgres.clj
+++ b/src/metabase/driver/postgres.clj
@@ -204,7 +204,7 @@
           :connection-details->spec  (u/drop-first-arg connection-details->spec)
           :date                      (u/drop-first-arg date)
           :prepare-value             (u/drop-first-arg prepare-value)
-          :set-timezone-sql          (constantly "UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';")
+          :set-timezone-sql          (constantly "SET local timezone to ?;")
           :string-length-fn          (constantly :CHAR_LENGTH)
           :unix-timestamp->timestamp (u/drop-first-arg unix-timestamp->timestamp)}))
```

Deployed and got this error:

> 2016-05-10T08:00:00.191459+00:00 app[web.1]: 05-10 08:00:00 DEBUG generic-sql.native :: SET local timezone to ?;
> 2016-05-10T08:00:00.194018+00:00 app[web.1]: 05-10 08:00:00 ERROR generic-sql.native :: Failed to set timezone: Batch entry 0 set local timezone to 'US/Pacific' was aborted.  Call getNextException to see the cause.

**Comment by andyparsons** [comment]
We also experienced this issue today after setting the timezone in the GUI.  It might be worth disabling that option in the next build or placing a warning on the page so others using Postgres don't get hung up by it. Great support here BTW.

**Comment by tlrobinson** [comment]
It looks like @agilliland is fixing this as part of #2564

**Comment by cvibhagool** [comment]
Hello guys! We just got started using Metabase, and it's been great.

Is this issue considered fixed, @tlrobinson @agilliland 
? I'm on Metabase 0.19.3, and got the "Failed to set timezone" error when running a query on a read-only 9.3 postgresql replica
```
10-05 09:34:48 DEBUG metabase.query-processor :: NATIVE FORM: 😳
{:query
 "SELECT CAST(\"public\".\"user_email_history\".\"created_at\" AS date) AS \"created_at\", count(*) AS \"count\" FROM \"public\".\"user_email_history\" WHERE CAST(\"public\".\"user_email_history\".\"created_at\" AS date) BETWEEN CAST((NOW() + INTERVAL '-31 day') AS date) AND CAST((NOW() + INTERVAL '-1 day') AS date) GROUP BY CAST(\"public\".\"user_email_history\".\"created_at\" AS date) ORDER BY CAST(\"public\".\"user_email_history\".\"created_at\" AS date) ASC",
 :params nil}


10-05 09:34:48 DEBUG generic-sql.query-processor :: ["UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';" "Asia/Bangkok"]

10-05 09:34:48 ERROR generic-sql.query-processor :: Failed to set timezone:
 BatchUpdateException:
 Message: Batch entry 0 UPDATE pg_settings SET setting = 'Asia/Bangkok' WHERE name ILIKE 'timezone' was aborted.  Call getNextException to see the cause.
 SQLState: 25006
 Error Code: 0
PSQLException:
 Message: ERROR: cannot execute SELECT in a read-only transaction
 SQLState: 25006
 Error Code: 0
```
Thanks

**Comment by camsaul** [comment]
Yeah.. #2564 wasn't really related to this, so not sure why it was closed. I'm going to reöpen this issue because the way we set timezones on Postgres (via a `SET` command) doesn't work on a read-only connection. We run SQL like this at the start of the transaction:

```sql
UPDATE pg_settings 
SET setting = 'US/Pacific' 
WHERE name ILIKE 'timezone'
```

Fortunately in #2984, which shipped as part of 0.19.0 I made some tweaks so the query will still run, albeït without setting the timezone at the beginning.

If anyone knows of some other way we can set the session timezone on a read-only Postgres connection let me know and we can get this working correctly 👍

**Comment by salsakran** [comment]
@camsaul  + @tlrobinson it looks like this is still an UPDATE in the current code base.

**Comment by camsaul** [comment]
@salsakran that's right see my comment from earlier today 😉

**Comment by salsakran** [comment]
oops. I had an open tab with this in the morning, and it seems to have never gotten updated. My bad.

**Comment by camsaul** [comment]
Worth reïterating, if anyone knows of some other way we can set the session timezone on a read-only Postgres connection chime in and I will implement it. I don't run around setting timezones on read-only connections too often so this is outside my realm of expertise 😉

**Comment by camsaul** [comment]
As @nickcarenza mentioned in #3730 using `SET SESSION TIME ZONE ...` to set the timezone should work on readonly connections.

**Comment by camsaul** [comment]
Almost exactly a year ago we were actually using SQL syntax like 

```sql
SET SESSION TIME ZONE TO ?
```

but that didn't work (it didn't like the timezone name being a prepared statement param)

So in #1431 we switched it to 

```sql
UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone'
```

To switch back to the `SET SESSION TIME ZONE TO` syntax we'll have to rework some things so timezone doesn't get passed as a prepared statement param

**Comment by camsaul** [comment]
Fixed by #4872
